<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Fun - Function manipulation.</title>
    <style>
        body { font-family: monospace; margin: 20px; line-height: 1.4; }
        a { color: #0066cc; }
        pre { white-space: pre-wrap; }
    </style>
</head>
<body>
    <div id="main-content">
<section class="p-strip is-bordered">
<div class="row">
<div class="col-3 u-hide--small u-hide" id="toc">
</div>
<div id="tableWrapper">
<p id="distroAndSection"></p>

Provided by: <a href="https://launchpad.net/ubuntu/questing/+package/ocaml-man">ocaml-man_5.3.0-3_all</a> <br><br><pre>
</pre><h4><b>NAME</b></h4><pre>
       Fun - Function manipulation.

</pre><h4><b>Module</b></h4><pre>
       Module   Fun

</pre><h4><b>Documentation</b></h4><pre>
       Module <b>Fun</b>
        : <b>sig</b> <b>end</b>

       Function manipulation.

       <b>Since</b> 4.08

   <b>Combinators</b>
       <u>val</u> <u>id</u> : <b>'a</b> <b>-&gt;</b> <b>'a</b>

       <b>id</b> is the identity function. For any argument <b>x</b> , <b>id</b> <b>x</b> is <b>x</b> .

       <u>val</u> <u>const</u> : <b>'a</b> <b>-&gt;</b> <b>'b</b> <b>-&gt;</b> <b>'a</b>

       <b>const</b> <b>c</b> is a function that always returns the value <b>c</b> . For any argument <b>x</b> , <b>(const</b> <b>c)</b> <b>x</b> is <b>c</b> .

       <u>val</u> <u>compose</u> : <b>('b</b> <b>-&gt;</b> <b>'c)</b> <b>-&gt;</b> <b>('a</b> <b>-&gt;</b> <b>'b)</b> <b>-&gt;</b> <b>'a</b> <b>-&gt;</b> <b>'c</b>

       <b>compose</b> <b>f</b> <b>g</b> is a function composition of applying <b>g</b> then <b>f</b> .  For any arguments <b>f</b> , <b>g</b> , and <b>x</b> , <b>compose</b> <b>f</b>
       <b>g</b> <b>x</b> is <b>f</b> <b>(g</b> <b>x)</b> .

       <b>Since</b> 5.2

       <u>val</u> <u>flip</u> : <b>('a</b> <b>-&gt;</b> <b>'b</b> <b>-&gt;</b> <b>'c)</b> <b>-&gt;</b> <b>'b</b> <b>-&gt;</b> <b>'a</b> <b>-&gt;</b> <b>'c</b>

       <b>flip</b> <b>f</b> reverses the argument order of the binary function <b>f</b> . For any arguments <b>x</b> and <b>y</b> , <b>(flip</b> <b>f)</b> <b>x</b> <b>y</b> is
       <b>f</b> <b>y</b> <b>x</b> .

       <u>val</u> <u>negate</u> : <b>('a</b> <b>-&gt;</b> <b>bool)</b> <b>-&gt;</b> <b>'a</b> <b>-&gt;</b> <b>bool</b>

       <b>negate</b> <b>p</b> is the negation of the predicate function <b>p</b> . For any argument <b>x</b> , <b>(negate</b> <b>p)</b> <b>x</b> is <b>not</b> <b>(p</b> <b>x)</b> .

   <b>Exception</b> <b>handling</b>
       <u>val</u> <u>protect</u> : <b>finally:(unit</b> <b>-&gt;</b> <b>unit)</b> <b>-&gt;</b> <b>(unit</b> <b>-&gt;</b> <b>'a)</b> <b>-&gt;</b> <b>'a</b>

       <b>protect</b>  <b>~finally</b>  <b>work</b>  invokes  <b>work</b> <b>()</b> and then <b>finally</b> <b>()</b> before <b>work</b> <b>()</b> returns with its value or an
       exception. In the latter case the exception is re-raised after <b>finally</b> <b>()</b>  .  If  <b>finally</b>  <b>()</b>  raises  an
       exception, then the exception <b>Fun.Finally_raised</b> is raised instead.

       <b>protect</b>  can be used to enforce local invariants whether <b>work</b> <b>()</b> returns normally or raises an exception.
       However, it does not protect against unexpected exceptions raised inside <b>finally</b> <b>()</b> such as <b>Out_of_memory</b>
       , <b>Stack_overflow</b> , or asynchronous exceptions raised by signal handlers (e.g.  <b>Sys.Break</b> ).

       Note: It is a programming error if other kinds of exceptions are raised by <b>finally</b>  ,  as  any  exception
       raised in <b>work</b> <b>()</b> will be lost in the event of a <b>Fun.Finally_raised</b> exception. Therefore, one should make
       sure to handle those inside the finally.

       <u>exception</u> <u>Finally_raised</u> <b>of</b> <b>exn</b>

       <b>Finally_raised</b>  <b>exn</b>  is  raised  by  <b>protect</b>  <b>~finally</b>  <b>work</b>  when <b>finally</b> raises an exception <b>exn</b> . This
       exception denotes either an unexpected exception or a programming error. As a general  rule,  one  should
       not catch a <b>Finally_raised</b> exception except as part of a catch-all handler.

OCamldoc                                           2025-06-12                                            <u><a href="../man3o/Fun.3o.html">Fun</a></u>(3o)
</pre>
 </div>
</div></section>
</div>
</body>
</html>